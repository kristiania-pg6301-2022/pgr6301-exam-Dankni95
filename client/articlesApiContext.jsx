import React from "react";
import { fetchJSON } from "./lib/fetchJSON";
import { postJSON } from "./lib/postJSON";

export const ArticlesApiContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },
  async listArticles(query) {
    return await fetchJSON("/api/articles?" + new URLSearchParams(query));
  },
  async createArticle(article) {
    return await postJSON("/api/articles", article);
  },
  async deleteArticle(query) {
    return await fetchJSON("/api/articles?" + new URLSearchParams(query));
  },
  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
  async endSession() {
    const res = await fetch("/api/login", { method: "DELETE" });
    if (!res.ok) {
      throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
    }
  },
});
