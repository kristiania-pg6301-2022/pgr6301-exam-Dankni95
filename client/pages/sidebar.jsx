import { Button, Grid } from "@mui/material";
import React from "react";

export default ({ user, socket }) => {
  function handleSubmitQuery(e) {
    socket.send(JSON.stringify({ category: e.target.value }));
  }

  return (
    <Grid container spacing={2}>
      <Grid item xs={2}>
        <Button
          onClick={(e) => {
            handleSubmitQuery(e);
          }}
          value=""
        >
          All
        </Button>
        <Grid item xs={8}>
          <Button
            value="Politics"
            onClick={(e) => {
              handleSubmitQuery(e);
            }}
          >
            Politics
          </Button>
        </Grid>
        <Button
          value="Economics"
          onClick={(e) => {
            handleSubmitQuery(e);
          }}
        >
          Economics
        </Button>
        <Grid item xs={8}>
          <Button
            value="Sport"
            onClick={(e) => {
              handleSubmitQuery(e);
            }}
          >
            Sport
          </Button>
        </Grid>
        <Button
          value="Health"
          onClick={(e) => {
            handleSubmitQuery(e);
          }}
        >
          Health
        </Button>
      </Grid>
    </Grid>
  );
};
