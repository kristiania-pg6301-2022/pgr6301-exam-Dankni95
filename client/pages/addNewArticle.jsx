import React, {useContext, useState} from "react";
import {useNavigate} from "react-router-dom";
import {FormInput, FormTextArea, FormTextOptions} from "../lib/formInput";
import {ArticlesApiContext} from "../articlesApiContext.jsx";

export function AddNewArticle({user}) {
    const {createArticle} = useContext(ArticlesApiContext);
    const [title, setTitle] = useState("");
    const [category, setCategory] = useState("Politics");
    const [content, setContent] = useState("");

    const navigate = useNavigate();

    function handleSubmit(event) {
        event.preventDefault();
        const {name} = user.openid
        createArticle({title, date: new Date().toLocaleDateString(), category, content, name});
        navigate("/");
    }

    return (
        <form onSubmit={handleSubmit}>
            <h1>Add new article</h1>

            <FormInput label={"Title:"} value={title} onChangeValue={setTitle}/>
            <FormTextOptions label={"Category: "} value={category} onChangeValue={setCategory}/>
            <FormTextArea label={"Content:"} value={content} onChangeValue={setContent}/>
            <div>
                <button>Submit</button>
            </div>
        </form>
    );
}
