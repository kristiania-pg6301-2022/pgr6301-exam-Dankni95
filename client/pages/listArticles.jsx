import React, {useContext, useState, useEffect} from "react";
import {useLoading} from "../useLoading";
import {ArticlesApiContext} from "../articlesApiContext.jsx";
import {useNavigate} from "react-router-dom";


export async function handleClick(article, deleteArticle) {
    await deleteArticle(article)
}


function Article({article: {title, date, category, content, name}, user}) {
    const {deleteArticle} = useContext(ArticlesApiContext)

    return (
        <div>
            <h6>{category}</h6>
            <h1>{"title" + title}</h1>
            <h6>{date}</h6>
            {user?.google || user?.openid ? <h4>{content}</h4> : "Sign in to read the article"}
            <h6>By: {name}</h6>
            {user.openid && user?.openid?.name === name ? <button onClick={(event) => handleClick( {del: title}, deleteArticle)}>Edit</button> : ""}
        </div>
    );
}


export function ListArticles({user}) {
    const {listArticles} = useContext(ArticlesApiContext);
    const [category, setCategory] = useState("");
    const [categoryQuery, setCategoryQuery] = useState("");
    const {loading, error, data} = useLoading(
        async () => await listArticles({category: category}),
        [category]
    );

    function handleSubmitQuery(e) {
        e.preventDefault();
        setCategory(categoryQuery);
    }

    if (loading) {
        return <div>Loading...</div>;
    }
    if (error) {
        return (
            <div>
                <h1>Error</h1>
                <div id="error-text">{error.toString()}</div>
            </div>
        );
    }

    return (
        <div>
            <h1>Articles</h1>

            <div>
                <form onSubmit={handleSubmitQuery}>
                    <label>
                        Category:
                        <select name={"Category"} value={categoryQuery}
                                onChange={(e) => setCategoryQuery(e.target.value)}
                        >
                            <option value="">All</option>
                            <option value="Politics">Politics</option>
                            <option value="Economics">Economics</option>
                            <option value="Sport">Sport</option>
                            <option value="Health">Health</option>
                        </select>
                        <button>Select</button>
                    </label>
                </form>
            </div>
            {data.map((article) => (
                <Article key={article.title} article={article} user={user} />
            ))}
        </div>
    );
}
