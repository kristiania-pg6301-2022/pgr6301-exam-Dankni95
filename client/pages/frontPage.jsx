import {Link, Navigate} from "react-router-dom"
import React from "react"
import {ListArticles} from "./listArticles";
import Sidebar from "./sidebar";


export function FrontPage({user}) {

    return (
        <>
            <div className="left-side"><Sidebar></Sidebar></div>
            <main className="main"><ListArticles user={user}/></main>
            <div className="right-side section yellow">Right Sidebar</div>
            <footer className="section coral">Footer</footer>
        </>
    )
}
