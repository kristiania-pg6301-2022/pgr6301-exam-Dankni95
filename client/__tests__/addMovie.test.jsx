import { AddNewArticle } from "../pages/addNewArticle";
import React from "react";
import ReactDOM from "react-dom";
import { Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";
import { ArticlesApiContext } from "../articlesApiContext.jsx";

describe("add movie component", () => {
  it("shows movies form", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <AddNewArticle />
      </MemoryRouter>,
      element
    );
    expect(element.innerHTML).toMatchSnapshot();
    expect(
      Array.from(element.querySelectorAll("form label strong")).map(
        (e) => e.innerHTML
      )
    ).toEqual(["Title:", "Year:", "Country:", "Plot:"]);
  });

  it("adds movie on submit", () => {
    const createMovie = jest.fn();
    const title = "Test movie";
    const element = document.createElement("div");
    ReactDOM.render(
      <ArticlesApiContext.Provider value={{ createMovie }}>
        <MemoryRouter>
          <AddNewArticle />
        </MemoryRouter>
      </ArticlesApiContext.Provider>,
      element
    );
    Simulate.change(element.querySelector(".form-input input"), {
      target: { value: title },
    });
    Simulate.change(element.querySelector(".form-input:nth-of-type(2) input"), {
      target: { value: "2022" },
    });
    Simulate.submit(element.querySelector("form"));
    expect(createMovie).toBeCalledWith({
      title,
      country: "",
      year: 2022,
      plot: "",
    });
  });
});
