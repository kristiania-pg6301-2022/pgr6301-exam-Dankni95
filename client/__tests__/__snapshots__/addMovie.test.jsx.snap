// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add movie component shows movies form 1`] = `"<form><h1>Add new movie</h1><div class=\\"form-input\\"><label><strong>Title:</strong> <input value=\\"\\"></label></div><div class=\\"form-input\\"><label><strong>Year:</strong> <input value=\\"\\"></label></div><div class=\\"form-input\\"><label><strong>Country:</strong> <input value=\\"\\"></label></div><div class=\\"form-input\\"><label><strong>Plot:</strong> <input value=\\"\\"></label></div><div><button>Submit</button></div></form>"`;
